src/Microsoft.AspNetCore.OData/Routing/ODataActionSelector.cs

```diff
@@ -126,31 +162,90 @@ public ActionDescriptor SelectBestCandidate(RouteContext context, IReadOnlyList<
                 // Method(key,relatedKey,ODataPath) vs Method(key,relatedKey).
                 var matchedCandidates = considerCandidates
-                    .Where(c => !c.FilteredParameters.Any() || TryMatch(c.FilteredParameters, availableKeys, optionalWrapper))
+                    .Where(c => TryMatch(context, c.FilteredParameters, availableKeys,
+                        optionalWrapper, c.TotalParameterCount, availableKeysCount))
                     .OrderByDescending(c => c.FilteredParameters.Count)
-                    .ThenByDescending(c => c.TotalParameterCount);
+                    .ThenByDescending(c => c.TotalParameterCount)
+                    .ToList();
+
+                // if there are still multiple candidate actions at this point, let's try some tie-breakers
+                if (matchedCandidates.Count > 1)
+                {
+                    // prioritize actions which explicitly declare the request method
+                    // e.g. using [AcceptVerbs("POST")], [HttpPost], etc.
+                    var bestCandidate = matchedCandidates.FirstOrDefault(candidate =>
+                        ActionAcceptsMethod(candidate.ActionDescriptor as ControllerActionDescriptor, context.HttpContext.Request.Method));
+                    if (bestCandidate != null)
+                    {
+                        return bestCandidate.ActionDescriptor;
+                    }
+                    
+                    // also priorize actions that have the exact number of parameters as available keys
+                    // this helps disambiguate between overloads of actions that implement actions
+                    // e.g. DoSomething(int key) vs DoSomething(), if there are no availableKeys, the
+                    // selector could still think that the `int key` param will come from the request body
+                    // and end up returning DoSomething(int key) instead of DoSomething()
+                    bestCandidate = matchedCandidates.FirstOrDefault(candidate => candidate.FilteredParameters.Count() == availableKeysCount);
+                    if (bestCandidate != null)
+                    {
+                        return bestCandidate.ActionDescriptor;
+                    }
+                }
 
-                // Return either the best matched candidate or the first
-                // candidate if none matched.
-                return (matchedCandidates.Any())
-                    ? candidates.Where(c => c.Id == matchedCandidates.FirstOrDefault().Id).FirstOrDefault()
-                    : candidates.FirstOrDefault();
+                return matchedCandidates.Select(c => c.ActionDescriptor).FirstOrDefault();
             }
 
             return _innerSelector.SelectBestCandidate(context, candidates);
         }
 
-        private bool TryMatch(IList<ParameterDescriptor> parameters, IList<string> availableKeys, ODataOptionalParameter optionalWrapper)
+        private bool TryMatch(
```

Comment:
Could you add a comment to showcase what is happening in this function as the number of params has increased.